ttnn/cpp/ttnn/tensor/tensor_impl.cpp

```diff
@@ -840,6 +847,214 @@ Tensor to_device<bfloat8_b>(
+            const auto num_shard_rows =
+                shard_height_idx == num_shards_height - 1 ? last_shard_height : logical_shard_shape.height();
+            const auto num_shard_cols =
+                shard_width_idx == num_shards_width - 1 ? last_shard_width : logical_shard_shape.width();
+
+            auto indices = LogicalPhysicalIdxPairs(num_shard_rows);
+            const auto logical_start_idx = shard_height_idx * logical_shard_shape.height() * logical_stride +
+                                           shard_width_idx * logical_shard_shape.width();
+            const auto physical_start_idx = shard_height_idx * physical_shard_shape.height() * physical_stride +
+                                            shard_width_idx * physical_shard_shape.width();
+            for (size_t i = 0; i < num_shard_rows; i++) {
+                indices[i] = {i * logical_stride + logical_start_idx, i * physical_stride + physical_start_idx};
+            }
+
+            logical_physical_mapping.push_back((LogicalPhysicalMapping){indices, num_shard_cols});
+        }
+    }
+    return logical_physical_mapping;
+};
+
+template <typename T>
+std::vector<T> convert_logical_data_to_physical_data(
+    const std::vector<T>& logical_data, const TensorSpec& tensor_spec) {
+    const auto& logical_shape = tensor_spec.logical_shape();
+    TT_FATAL(
+        logical_data.size() == logical_shape.volume(),
+        "Logical data size {} should be same as volume indicated by logical shape {}",
+        logical_data.size(),
+        logical_shape);
+
+    const auto& physical_shape = tensor_spec.physical_shape();
+    auto [logical_shard_shape, physical_shard_shape] = get_logical_and_physical_shard_shapes(tensor_spec);
+
+    std::vector<T> physical_data(physical_shape.height() * physical_shape.width(), 0);
+
+    auto logical_2D_shape = flatten_to_2D(logical_shape);
+    size_t physical_stride = physical_shape.width();
+
+    const auto logical_physical_mapping = compute_logical_to_physical_shards_mapping(
+        logical_2D_shape, logical_shard_shape, physical_shard_shape, physical_stride);
+
+    for (const auto& [indices, cols] : logical_physical_mapping) {
+        for (const auto& idx_pair : indices) {
+            auto logical_idx_start = idx_pair[0];
+            auto physical_idx_start = idx_pair[1];
+
+            for (size_t col = 0; col < cols; col++) {
+                physical_data[physical_idx_start + col] = logical_data[logical_idx_start + col];
+            }
+        }
+    }
+
+    TT_FATAL(
+        physical_data.size() == physical_shape.height() * physical_shape.width(),
+        "Physical data size {} should be same as volume indicated by physical shape {}",
+        physical_data.size(),
+        physical_shape);
+
+    if (tensor_spec.layout() == Layout::TILE) {
+        // TODO: Fix convert_layout_row_major_to_tile to take in vector instead of buffer?
```

Comment:
I saw this comment a couple times.. what do we want to do here? did you intend to add an impl for this?

Reply:
Need to review our data containers that gets passed around these functions. `convert_layout_row_major_to_tile` and `convert_layout_tile_to_row_major` takes buffer and returns vector, whereas I prefer just vectors everywhere. I think the assumption was that those functions were the only ones you needed to call, so just passing original buffer from a tensor made sense.